All the code

rosskarchner opened the pull request:
There's a few rough edges and missing features:
- it will silently overwrites existing build and release zip files (if you re-use a build label, for example)
- There's no code in place that supports a persistent 'lib' directory (for apps and modules that will be deployed on their own schedule, separate from the project release cycle)
- setup.py needs love
- the --append-wsgi option was built to support newrelic, but it turns out the newrelic code needs _pre_-pended.
- "paths.json" is thus far undocumented, but it should be pretty easy to figure out if you look at dfd_paths.py

... but I think this is a good conversation starter ;)

rosskarchner committed:
Initial commit!

rosskarchner committed:
Merge branch 'master' of github.com:rosskarchner/drama-free-django

rosskarchner committed:
move README template to rst

rosskarchner committed:
Started writing README

rosskarchner committed:
Update __init__.py

rosskarchner committed:
Replaced PBR with a simplified setup.py-- this needs further work.

rosskarchner committed:
Added "make a build"

rosskarchner committed:
Restore missing console_script

rosskarchner committed:
Merge branch 'master' of github.com:rosskarchner/drama-free-django

rosskarchner committed:
Update README.rst

rosskarchner committed:
add missing extended_wsgi

rosskarchner committed:
improved logic for locating zipped wheel dir

rosskarchner committed:
improved logic for locating zipped wheel dir

rosskarchner committed:
have wsgi.py import sitecustomize

Seems like it doesn't happen automatically under mod_wsgi

rosskarchner committed:
expanded "make a build" section

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

rosskarchner committed:
Update README.rst

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+                find_links=None):
+    call_list = ["pip", "wheel", "--wheel-dir=%s" % destination]
+
+    if packages is not None:
```
use truthiness, i.e.

```
if not packages:
```

willbarton replied:
That will only check `packages.__nonzero__`. 

I like the more explicit `packages is not None`. It's clear exactly what the case is.

vickumar1981 replied:
the line above will check for not none.  it is extraneous in python to do a check for not None.  it's simpler and cleaner the way i presented above.

willbarton replied:
It will check for `not None`, as well as other possible cases that fail a boolean where we'd either want to handle it differently or know why `packages` was a completely unexpected type. The potential for unintended side-effects suggests to me that leaving out the explicit `is [not] None` in all of these cases is not simpler or cleaner.

[PEP8 also warns against test `if foo` as opposed to `if foo is not None`.](https://www.python.org/dev/peps/pep-0008/#programming-recommendations)

vickumar1981 replied:
in this case, i think we expect packages to be a list.  so you don't really have to worry about the boolean value of False here.

In fact, if you do pass in a boolean, the above code will still fail.

You can't use the += operator with a list and a boolean.

That operator tells me packages HAS to be a list.

willbarton replied:
In this case, I'd suggest `packages` default to `[]`, and then there's no need for the check (and if it's not a list, an appropriate exception will be raised).

But generally, I'd recommend strongly against `if foo` in place of `if foo is not None`.

vickumar1981 replied:
https://www.python.org/dev/peps/pep-0008/#programming-recommendations

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+    if packages is not None:
+        call_list += packages
+
+    if find_links is not None:
```
ditto.

also, you can re-write this as an if-comprehension:

call_list += ['--find-links=%s' % find_links] if find_links else list()

rosskarchner replied:
the "find links" stuff is not longer used, so I've cut it out.

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+    if find_links is not None:
+        call_list += ['--find-links=%s' % find_links]
+
+    if requirements_file is not None:
```
same.

rosskarchner replied:
'Writing Idiomatic Python' recommends strict None comparisons when checking function arguments that default to None. (but trusting truthiness otherwise)

In most of these cases, the difference seems rather academic, though. Seems like maybe we need to start work on a style guide ;)

willbarton replied:
Open an issue in https://github.com/cfpb/back-end! :grin:

[Google's Python Style Guide](https://google.github.io/styleguide/pyguide.html#True/False_evaluations) also mentions this as an exception to trusting implicit falseness, for reasons of ambiguity.

vickumar1981 replied:
Right.  That's 100% correct.

But, the function arguments don't NEED to default to None.

```
def save_wheels(destination, packages=None,
    requirements_file=None,
    find_links=None):
   ...
```

That's the function as is.  But, let's say I rewrote that to simply:

```
def save_wheels(destination, packages=list(),
    requirements_file='',
    find_links=False):
```

Now, I'm being clear about what arguments I expect to my function, and I don't have to use the is None is Not None comparisons according to both PEP8 and 'Writing Idiomatic Python', I can use truthiness b/c I didn't rely on passing in a None value into my function.  It makes the code easier to comprehend for person xyz 10 years from now.

vickumar1981 replied:
There is an exception to this that I think it is important to understand.

If packages or requirements were mutable inside the save_wheels function, then it makes sense to default them to None.  But, in this case they are not changed within that function scope, which is why defaulting them to None doesn't make sense.

vickumar1981 commented on no_drama/build_lib/dfd_settings.py:
```diff
@@ -0,0 +1,20 @@
+
+real_settings = os.environ.get('REAL_DJANGO_SETTINGS')
+
+if real_settings is not None:
```
truthiness.

rosskarchner replied:
agreed, I'll make that change

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def get_template(filename, mode='r'):
+    template_file = get_file_from_dir(templates_dir, filename, mode=mode)
```
what happens if the file doesn't exist?  i feel like get_file_from_dir should handle that exception.

rosskarchner replied:
I think failing-with-an-exception probably IS the right thing-- these are files that are bundled with no-drama itself.

vickumar1981 replied:
i would just wrap that line with a block:

```
fname = get_file_from_dir(...)
if os.path.isfile(fname):
   ...
else:
  raise SomeCustomException
```

either that, or you should have this in a try..except.

willbarton replied:
I think it'd make more sense to `try`/`except` wherever the error should be appropriately handled.

rosskarchner replied:
I'm factoring these two functions out-- an earlier version used templates everywhere, and now there's only one. I think I'll be able to convert that to a plain old file soon, too.

vickumar1981 commented on no_drama/build_lib/dfd_paths.py:
```diff
@@ -0,0 +1,37 @@
+        joined = os.path.join(root, relpath)
+        return os.path.normpath(joined)
+    else:
+        return None
```
don't need that else block.  if you don't return anything from a python function, None is the default return value, i believe, but it's a bad idea to return None's or null's from a function, or to pass them in as parameters.

rosskarchner replied:
I wanted to be explicit about returning None here-- someone eyeballing the function should be able to know that None is an expected return-- I don't think it's obvious, without the else.

I'd like to know more about returning None being bad form, though. In this case, the obvious alternative would be empty strings, which seems... sloppier, to me.

vickumar1981 replied:
I think Robert Martin talks about this with nulls in java.  The problem is it is unclear what none signals.  Is it an error?  Does it just mean that it can be safely ignored?  Different programmers use none and null in different contexts to mean different things causing ambiguity.

This is one reason you'll see the error none type has no attribute...or a null pointer exception in java.

vickumar1981 commented on no_drama/build_lib/activate_phase2.py:
```diff
@@ -0,0 +1,8 @@
+
+from dfd_paths import root, update_symlink
+
+if update_symlink is not None:
```
same truthiness.

rosskarchner committed:
Update README.rst

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+    build_parser.set_defaults(func=stage_bundle)
+
+    # release parser arguments
+    release_parser.add_argument('build_zip', help='path to a zip file'
```
You could create a dictionary and iterate it and call add arguments in a loop, reducing the lines of code.

rosskarchner committed:
default to packages=[] in save_wheels

Thanks for the feedback @vickumar1981 and @willbarton

rosskarchner committed:
nix the find_links functionality from save_wheels

It's a leftover from when the app was maintaining it's own (pointless)
cache of wheels

rosskarchner committed:
the truth(iness) shall set you free!

rosskarchner committed:
Removed some functions that only exist for one file

vickumar1981 commented on no_drama/__main__.py:
```diff
@@ -0,0 +1,179 @@
+    bootstrap_wheels_destination = os.path.join(build_dir,'bootstrap_wheels')
+    save_wheels(packages=bootstrap_wheels, destination=bootstrap_wheels_destination)
+
+    # move just the wheels we want into the bundle dir  
```
i think this method can be broken down a bit into smaller sub-routines.

each commented block really should just be it's own function:

```
    def activate_wheels(...)
    def move_wheels_to_bundle_dir(...)
    def copy_django_proj_to_bundle_dir(...)
    def make_lib_directory(...)
    def generate_activate_shell_file(...)
```

then you can just get rid of the comments themselves.

rosskarchner replied:
That's a good call

rosskarchner committed:
Switched from None comparison to truthiness check.

rosskarchner committed:
you are

rosskarchner committed:
not-quite done yet: usese a skeleton directory, instead
of copying things one at a time, no longer needs templates

rosskarchner committed:
root all commands to the directory where activate.sh lives

rosskarchner committed:
properly escape $dir

rosskarchner committed:
one last dollar sign

rosskarchner committed:
Dollar blindness, apparently

rosskarchner committed:
Update activate.sh.template

rosskarchner committed:
use wheel --force to install

This means, we don't protect against incompatibilities between build server and deployment server. 

It seems like the differences CAN be minor though. I'm still investigating

rosskarchner committed:
Update activate.sh.template

rosskarchner committed:
add dependencies to pip, wheel, setuptools

rosskarchner committed:
Merge branch 'skel-refactor'

rosskarchner committed:
Merge branch 'master' of github.com:rosskarchner/drama-free-django

rosskarchner committed:
Update __main__.py

rosskarchner committed:
added a readme to paths.d

rosskarchner committed:
README, nm

rosskarchner committed:
Update MANIFEST.in

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update dfd_settings.py

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update dfd_settings.py

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update dfd_settings.py

rosskarchner committed:
Merge branch 'master' of github.com:rosskarchner/drama-free-django

rosskarchner committed:
new --aux argument for adding extra files

rosskarchner committed:
fix indentation error

rosskarchner committed:
delete the staging directory created during release

rosskarchner committed:
Builds are now executable!

rosskarchner committed:
Merge branch 'master' of github.com:rosskarchner/drama-free-django

rosskarchner committed:
bring back the wheel cache

rosskarchner committed:
Nevermind

Need to rethink wheel caching

rosskarchner committed:
you can now specify multiple requirements files, and the requirements are cached!

rosskarchner committed:
fix for requirements caching

rosskarchner committed:
remove PDB

rosskarchner committed:
caches wheels as a giant blob-- not ideal, but should make redundant builds faster

rosskarchner committed:
adds a --static option at build time to inject static files.

- a default directory for these extra static files
- support in dfd_settings for the additional static files

rosskarchner committed:
add a path for debug_if_exists

rosskarchner committed:
add settings support for debug_if_exists

rosskarchner committed:
add path to secret key file, and get_path_if_exists

rosskarchner committed:
add secret_key support to settings

rosskarchner committed:
builds, when run as an executable, now support a --destination / -d arg...

... which lets you tell it where to deploy to.

rosskarchner committed:
finally doing wheel caching right

I think

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
factor some logic to logical modules

rosskarchner committed:
Update setup.py

rosskarchner committed:
Update __main__.py

rosskarchner committed:
Update sitecustomize.py

rosskarchner committed:
Update sitecustomize.py

rosskarchner committed:
Update activate.sh

rosskarchner committed:
Update activate.sh

rosskarchner committed:
fix a bug in wsgi-prepend logic

willbarton commented:
Should this get merged since we appear to actually be using this? :)

rosskarchner commented:
I'm going to merge now-- we should make sure deployments use the 'cfpb' version from now on
